vignettes/creating_linters.Rmd

```diff
@@ -11,113 +11,186 @@ vignette: >
+object that parses the input file in two ways -- once is done expression-by-expression, the other contains all of the
+expressions in the file. This is done to facilitate caching. Suppose your package has a long source file (e.g., 100s of expressions) --
+rather than run linters on every expression every time the file is updated, when caching is activated `lintr` will
+only run the linter again on expressions that have changed.
+
+Therefore, it is preferable to write expression-level linters whenever possible. Two types of exceptions observed in `lintr` are
+(1) when several or all expressions are _required_ to ensure the linter logic applies (e.g., `conjunct_test_linter` looks for
+consecutive calls to `stopifnot()`, which will typically appear on consecutive expressions) or (2) when the linter logic
+is very simple & fast to compute, so that the overhead of re-running the linter is low (e.g., `single_quotes_linter`).
 
 ```r
-Lint(
-  filename = source_file$filename,
-  line_number = parsed$line1,
-  column_number = parsed$col1,
-  type = "style",
-  message = "Use <-, not =, for assignment.",
-  line = source_file$lines[parsed$line1]
+xml <- source_expression$xml_parsed_content
+
+xpath <- "//expr[preceding-sibling::SPECIAL[text() = '%>%'] and *[1][self::SYMBOL]]"
+bad_expr <- xml2::xml_find_all(xml, xpath)
+```
+
+Here is the core linter logic. `source_expression$xml_parsed_content` is copied to a local variable
+(this is not strictly necessary but facilitates debugging). `xpath` is an XPath expression
+for expressions matching the discouraged usage. `bad_expr` is the result of executing this logic
+on the expression's XML representation.
+
+Let's examine the XPath a bit more closely:
+
+```xpath
+//expr                  # global search for 'expr' nodes (R expressions), at any nesting level
+[                       # node[...] looks for any 'node' satisfying conditions in ...
+  preceding-sibling::   # "siblings" are at the same nesting level in XML
+    SPECIAL[            # 'SPECIAL' is the parse token for infix operators like %% or %+%
+      text() = '%>%'    # text() returns the string associated with this node
+    ]                   #
+  and                   # combine conditions with 'and'
+  *                     # match any node
+  [1]                   # match the first such node
+  [self::SYMBOL]        # match if the current node is a 'SYMBOL' (i.e., a 'name' in R)
+]                       #
+```
+
+Taken together, that means we want to match `expr` nodes preceded by the `%>%` infix operator whose first child node is a `name`.
+That maps pretty closely to the description of what the `pipe_call_linter` is looking for, but there is subtlety in mapping
+between the R code you're used to and how they show up in the XML representation. `expr` nodes in particular take some practice
+to get accustomed to -- use the plentiful XPath-based linters in `lintr` as a guide to get extra practice.
```

Comment:
Maybe also reference the excellent w3schools tutorials here again? Also worth noting that `xml2` implements XPath 1.0.